background/services/provider-bridge/index.ts

```diff
@@ -51,36 +58,108 @@ export default class ProviderBridgeService extends BaseService<Events> {
-          logger.log(
-            `background: request payload: ${JSON.stringify(event.request)}`
-          )
-          const response = {
-            id: event.id,
-            result: await this.routeContentScriptRPCRequest(
-              event.request.method,
-              event.request.params
-            ),
-          }
-          logger.log("background response:", response)
-
-          port.postMessage(response)
-        })
+      if (port.name === EXTERNAL_PORT_NAME && port.sender?.url) {
+        const listener = this.onMessageListener(
+          port as Required<browser.Runtime.Port>
+        )
+        port.onMessage.addListener(listener)
+        // TODO: store port with listener to handle cleanup
       }
     })
 
     // TODO: on internal provider handlers connect, disconnect, account change, network change
   }
 
-  routeContentScriptRPCRequest(
+  onMessageListener(
+    port: Required<browser.Runtime.Port>
+  ): (event: PortRequestEvent) => Promise<void> {
+    const url = port.sender.url as string
+    const favIconUrl = port.sender.tab?.favIconUrl ?? ""
+
+    return async (event: PortRequestEvent) => {
+      // a port: browser.Runtime.Port is passed into this function as a 2nd argument by the port.onMessage.addEventListener.
+      // This contradicts the MDN documentation so better not to rely on it.
+      logger.log(
+        `background: request payload: ${JSON.stringify(event.request)}`
+      )
+
+      const response: PortResponseEvent = { id: event.id, result: [] }
+
+      if (
+        event.request.method === "eth_requestAccounts" &&
+        !(await this.checkPermission(url))
+      ) {
+        const permissionRequest: PermissionRequest = {
+          url,
+          favIconUrl,
+          state: "request",
+        }
+
+        const blockUntilUserAction = await this.requestPermission(
+          permissionRequest
+        )
+        await blockUntilUserAction
+
+        if (await this.checkPermission(url)) {
+          response.result = new UserRejectedRequestError()
+        }
+      }
+
+      if (await this.checkPermission(url)) {
+        response.result = await this.routeContentScriptRPCRequest(
+          event.request.method,
+          event.request.params
+        )
+      } else {
+        response.result = new UnauthorizedError()
+      }
+      logger.log("background response:", response)
+
+      port.postMessage(response)
+    }
+  }
+
+  async requestPermission(permissionRequest: PermissionRequest) {
+    let blockResolve: (value: unknown) => void | undefined
+    const blockUntilUserAction = new Promise((resolve) => {
+      blockResolve = resolve
+    })
+
+    this.emitter.emit("permissionRequest", permissionRequest)
+    await ProviderBridgeService.showDappConnectWindow()
+
+    // ts compiler does not know that we assign value to blockResolve so we need to tell him
+    this.#pendingPermissionsRequests[permissionRequest.url] = blockResolve!
+    return blockUntilUserAction
```

Comment:
This is working accidentally I believe. The promise constructor isn't guaranteed to run synchronously, and it's only because of the intervening `await` that this gets set correctly. We should set this inside the promise function and skip the intermediate variable. 

```suggestion
    this.emitter.emit("permissionRequest", permissionRequest)
    ProviderBridgeService.showDappConnectWindow()

    // This promise will be resolved by a call to grantPermission
    // or denyOrRevokePermission.
    return new Promise((resolve) => {
      this.#pendingPermissionsRequests[permissionRequest.url] = resolve
    })
```

I think this is overall clearer.